### PR TITLE
WIP: Changes for OpenVR 1.0.7

### PR DESCRIPTION
**Pull request by Ralith**
This gets the library building against the proposed -sys update and adds support for some important additional functionality such as controller input events. However, it is incomplete; it derives from what was originally intended to be a ground-up rewrite, and focuses on the bare minimum necessary to get up and running.

Substantial refactoring--some necessary, some cleanup, and some cosmetic--is included, again as a result of the original conception as a rewrite.

### CODE REVIEW

**Comment by palodequeso** [comment]
Looks like this depends on Vulkan now? I'm not a maintainer/reviewer, but I was just reading through. I plan to make a VR game with Amethyst and this at some point. Note: I understand that you pretty much need a Vulkan compatible device to run VR anyway, I'm just curious.

File: src/lib.rs [hunk 1]
```diff
@@ -1,62 +1,143 @@
...
+    Ok(unsafe { Context::new() }?)
+}
+
+pub struct System<'a>(&'a sys::VR_IVRSystem_FnTable);
+pub struct Compositor<'a>(&'a sys::VR_IVRCompositor_FnTable);
+
+/// Entry points into OpenVR.
+///
+/// At most one of this object may exist at a time.
+pub struct Context {
```
Comment (reneeichhorn) [inline src/lib.rs]: What I don't like on this approach is that you are forcing the user to instantiate all interfaces / subsystems.
If one of them fails the whole Context will be invalid. I think depending on the ApplicationType some interfaces are not available in will cause an error. The interfaces should be independent from each other.
As a quick solution we'd at least wrap an Option around each interface and make them all optional.

**Comment by reneeichhorn** [comment]
@palodequeso The vulkan code you mentioned is actually part of OpenVR itself.
See https://github.com/ValveSoftware/openvr/wiki/Vulkan

File: src/compositor.rs [hunk 2]
```diff
@@ -1,73 +1,199 @@
...
+    ///
+    /// If `bounds` is None, the entire texture will be used. Lens distortion is handled by the OpenVR implementation.
+    pub fn submit(&self, eye: Eye, texture: &Texture, bounds: Option<&TextureBounds>) -> Result<(), CompositorError> {
+        use self::TextureHandle::*;
+        let texture = sys::Texture_t {
+            handle: match texture.handle {
+                Vulkan(ref x) => x as *const _ as *mut _,
+            },
+            eType: match texture.handle {
+                Vulkan(_) => sys::ETextureType_ETextureType_TextureType_Vulkan,
```
Comment (reneeichhorn) [inline src/compositor.rs]: Reminder: 
ToDo: OpenGL support

**Comment by palodequeso** [comment]
Oh rad, that makes sense, since it does the eye compositing IIRC.

File: src/compositor.rs [hunk 3]
```diff
@@ -1,73 +1,199 @@
...
+    pub game: TrackedDevicePoses,
+}
+
+pub use sys::VkPhysicalDevice_T;
+pub use sys::VkDevice_T;
+pub use sys::VkInstance_T;
+pub use sys::VkQueue_T;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Texture {
```
Comment (reneeichhorn) [inline src/compositor.rs]: Reminder:
ToDo: I'd prefer it to split the compositor into a module with multiple submodules for texture, etc.
```
+ compositor
    + mod.rs <- general logic
    + texture<- texture related stuff
           + mod.rs
           + vulkan.rs
           + opengl.rs
    + error.rs <- error enum ?maybe not?
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,62 +1,143 @@
...
+}
+
+impl fmt::Display for InitError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad(error::Error::description(self))
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Eye {
```
Comment (reneeichhorn) [inline src/lib.rs]: I actually like that they were in the common module since they are also used in some other interfaces.

File: src/system.rs [hunk 5]
```diff
@@ -1,112 +1,587 @@
...
+            sys::EVREventType_EVREventType_VREvent_PerformanceTest_FidelityLevel => PerformanceTest_FidelityLevel,
+            sys::EVREventType_EVREventType_VREvent_MessageOverlay_Closed => MessageOverlay_Closed,
+            x if x >= sys::EVREventType_EVREventType_VREvent_VendorSpecific_Reserved_Start
+                && x <= sys::EVREventType_EVREventType_VREvent_VendorSpecific_Reserved_End => VendorSpecific(x),
+            x => Unknown(x),
         }
     }
 }
+
+pub type PropertyContainerHandle = sys::PropertyContainerHandle_t;
```
Comment (reneeichhorn) [inline src/system.rs]: We can definitely split this up

**Comment by Ralith** [comment]
To be clear, Vulkan is not, in principle, a hard dep of OpenVR; the new code just does not expose the OpenGL/DX paths.

File: src/lib.rs (hunk 4 above)
Comment (Ralith) [inline src/lib.rs]: Since these are in the root, they can be used from any module in the crate.

**Comment by Ralith** [comment]
I split up the modules some more and drafted OpenGL support. I adjusted Context not to eagerly load subsystems; now it merely serves as a zero-cost safety mechanism.

As far as I know, github doesn't have any ability to target a PR at a nonexistent branch. If you'd like to create one, that might be helpful to allow this to be merged early and permit collaboration towards updating the remaining components.

**Comment by reneeichhorn** [comment]
I changed the target branch to "development" (branched of from master)

**Comment by reneeichhorn** [comment]
Merged it now into the development branch for further improvements and testing

**Pull request merged**